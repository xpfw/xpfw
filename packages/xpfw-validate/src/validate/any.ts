import { get, intersection, isFunction, isNil, isNumber, isObject, isString, keys, set } from "lodash"
import { ErrorType, FieldType, Method } from "../globals"
import { IErrorBase, ValidateFunction } from "../typeDef"
import validateArray from "./array"
import validateBool from "./bool"
import validateDate from "./date"
import validateLocation from "./location"
import validateNumber from "./number"
import validateObject from "./object"
import validateQueryObject, { possibleQueryOperators } from "./queryObject"
import validateSelect from "./select"
import validateText from "./text"
import checkShouldBeRequired from "./util/checkShouldBeRequired"

const validateAny: ValidateFunction = (value, field, params) => {
  let errorResolve: any = null
  if (get(params, "method") === Method.Find && isObject(value)) {
    if (intersection(keys(value), possibleQueryOperators).length > 0) {
      const form: any = params.form
      return validateQueryObject(value, form, params)
    }
  }
  if (isObject(field) && !isNil(field.validate)) {
    if (isNumber(field.type)) {
      switch (field.type) {
        case FieldType.Text:
        case FieldType.Password:
        case FieldType.RelationshipSingle: {
          errorResolve = validateText(value, field, params)
          break
        }
        case FieldType.Number:
        case FieldType.Slider: {
          errorResolve = validateNumber(value, field, params)
          break
        }
        case FieldType.Boolean: {
          errorResolve = validateBool(value, field, params)
          break
        }
        case FieldType.Date: {
          errorResolve = validateDate(value, field, params)
          break
        }
        case FieldType.Select: {
          errorResolve = validateSelect(value, field, params)
          break
        }
        case FieldType.Object: {
          errorResolve = validateObject(value, field, params)
          break
        }
        case FieldType.Array:
        case FieldType.RelationshipMulti: {
          if (field.type === FieldType.RelationshipMulti) {
            set(field, "validate.type", FieldType.Text)
          }
          errorResolve = validateArray(value, field, params)
          break
        }
        case FieldType.Location: {
          errorResolve = validateLocation(value, field, params)
          break
        }
        default: {
          errorResolve = null
          break
        }
      }
    } else if (isString(field.type)) {
      if (isObject(params) && isObject(params.registry)) {
        const validator = params.registry.findValidator(field.type)
        if (isFunction(validator)) {
          errorResolve = validator(value, field, params)
        }
      }
    }
    if (isNil(errorResolve)) {
      console.log("erroirrseolve is null for", field)
      errorResolve = Promise.reject(false)
    }
  } else {
    errorResolve = Promise.resolve(value)
  }
  if (isObject(field) && isObject(field.validate)) {
    const validate: any = field.validate
    if (isFunction(validate.custom)) {
      let catched = false
      const errs: any = {mapTo: field.mapTo, errors: []}
      return errorResolve.catch((e: any) => {
        catched = true
        errs.errors = e
        return validate.custom(value, field, params)
      }).then(() => {
        if (!catched) {
          return validate.custom(value, field, params)
        }
        return Promise.reject(errs)
      })
    }
  }
  return errorResolve
}

export default validateAny
