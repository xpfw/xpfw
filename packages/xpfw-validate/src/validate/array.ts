import { cloneDeep, each, get, isArray, isNil, isNumber, isString } from "lodash"
import { ErrorType, FieldType } from "../globals"
import { IErrorBase, ValidateFunction } from "../typeDef"
import validateAny from "./any"
import checkShouldBeRequired from "./util/checkShouldBeRequired"

const validateArray: ValidateFunction = (value, field, params) => {
  const errors: IErrorBase[] = []
  const validArray = isArray(value)
  if (checkShouldBeRequired(field, params) && !validArray) {
    errors.push({
      type: ErrorType.required
    })
  }
  if (validArray) {
    const min = get(field, "validate.min")
    if (isNumber(min)) {
      if (value.length < min) {
        errors.push({
          type: ErrorType.tooLow
        })
      }
    }
    const max = get(field, "validate.max")
    if (isNumber(max)) {
      if (value.length > max) {
        errors.push({
          type: ErrorType.tooHigh
        })
      }
    }
  }
  if (errors.length > 0) {
    return Promise.reject(errors)
  }
  const promises: Array<Promise<any>> = []
  if (!isNil(field.validate)) {
    const subValidate: any = field.validate
    if (isString(subValidate.type) || isNumber(subValidate.type)) {
      each(value, (date) => {
        const tmpFieldOptions = cloneDeep(field)
        tmpFieldOptions.type = subValidate.type
        if (tmpFieldOptions.type === FieldType.Array || tmpFieldOptions.type === FieldType.Object) {
          tmpFieldOptions.validate = subValidate.validate
        }
        promises.push(validateAny(date, tmpFieldOptions, params))
      })
    }
  }
  return Promise.all(promises).then((results) => {
    return value
  })
}

export default validateArray
