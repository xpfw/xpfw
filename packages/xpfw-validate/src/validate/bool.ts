import { isBoolean, isNil } from "lodash"
import { ErrorType } from "../globals"
import { IErrorBase, ValidateFunction } from "../typeDef"
import checkShouldBeRequired from "./util/checkShouldBeRequired"

const validateBool: ValidateFunction = (value, field, params) => {
  const errors: IErrorBase[] = []
  const isValidBoolean = isBoolean(value)
  if (checkShouldBeRequired(field, params) && (!isValidBoolean || !value)) {
    errors.push({
      type: ErrorType.required
    })
  }
  if (!isNil(value) && !isValidBoolean) {
    errors.push({
      type: ErrorType.invalidType
    })
  }
  if (errors.length > 0) {
    return Promise.reject({mapTo: field.mapTo, errors})
  }
  return Promise.resolve(value)
}

export default validateBool
