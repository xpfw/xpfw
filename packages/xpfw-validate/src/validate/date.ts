import { isDate, isNil, isString } from "lodash"
import * as moment from "moment"
import { ErrorType } from "../globals"
import { IErrorBase, ValidateFunction } from "../typeDef"
import checkShouldBeRequired from "./util/checkShouldBeRequired"
const m: any = moment
const validateDate: ValidateFunction = (value, field, params) => {
  if (isString(value) && m(value).isValid()) {
    value = m(value).toDate()
  }
  const errors: IErrorBase[] = []
  const isValidDate = isDate(value)
  if (checkShouldBeRequired(field, params) && !isValidDate) {
    errors.push({
      type: ErrorType.required
    })
  }
  if (!isNil(value) && !isValidDate) {
    errors.push({
      type: ErrorType.invalidType
    })
  }
  if (errors.length > 0) {
    return Promise.reject({mapTo: field.mapTo, errors})
  }
  return Promise.resolve(value)
}

export default validateDate
