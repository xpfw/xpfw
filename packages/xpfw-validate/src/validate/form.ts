import { get, indexOf, intersection, isNil, isObject, isString, keys, set } from "lodash"
import { ErrorType, Method } from "../globals"
import getPermittedFields from "../permission/getPermittedFields"
import { IForm, IFormError, IParametersBase } from "../typeDef"
import validateField from "./any"
import validateQueryObject from "./queryObject"
import checkForFieldInSubKeys from "./util/checkForFieldInSubKeys"
import getRequireOwner from "./util/getRequireOwner"

const updateArrayOperators: string[] = [
  "$", "$inc", "addToSet", "$pop", "$pullAll", "$pull", "$pushAll", "$push",
  "$each", "$slice", "$sort", "$position", "$inc", "$mul", "$rename", "$setOnInsert",
  "$set", "$unset", "$min", "$max", "$currentDate", "$limit", "$skip", "$or", "$and"
]

const validateForm: (value: any, form: IForm, params: IParametersBase) => Promise<any> =
async (value, form, params) => {
  if (isObject(params)) {
    if (isNil(params.form)) {
      params.form = form
    }
    switch (params.method) {
      case Method.Find: {
        return validateQueryObject(value, form, params)
        break
      }
      case Method.Update: {
        try {
          const newRes = await checkForFieldInSubKeys(value, form, params, updateArrayOperators)
          return newRes
        } catch (e) {
          if (e !== -1) {
            throw e
          }
        }
        break
      }
    }
  }
  const validatedObject = {}
  const fields = await getPermittedFields(form, params, true)
  if (fields.length === 0) {
      return Promise.resolve(validatedObject)
  }
  const formError: IFormError = {
    model: form.model,
    errors: []
  }

  for (const field of fields) {
    try {
      const newVal = await validateField(get(value, field.mapTo), field, params)
      if (!isNil(newVal)) {
        set(validatedObject, field.mapTo, newVal)
      }
    } catch (e) {
      formError.errors.push(e)
    }
  }
  if (formError.errors.length > 0) {
    return Promise.reject(formError)
  }
  const method = isObject(params) && isString(params.method) ? params.method : ``
  if (method === Method.Update && keys(validatedObject).length === 0) {
    return Promise.reject({model: form.model, errors: [{
      model: `none`, type: ErrorType.emptyUpdate
    }]})
  }
  const requireOwner = getRequireOwner(form, method)
  if (isString(requireOwner)) {
    const idPath: any = params.idPath
    set(validatedObject, requireOwner, get(params.currentUser, idPath))
  }
  return Promise.resolve(validatedObject)
}

export default validateForm
