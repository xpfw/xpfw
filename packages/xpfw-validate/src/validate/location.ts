import { isNaN, isNil, isNumber, isString } from "lodash"
import { ErrorType } from "../globals"
import { IErrorBase, ValidateFunction } from "../typeDef"
import checkShouldBeRequired from "./util/checkShouldBeRequired"

const validateLocation: ValidateFunction = (value, field, params) => {
  const errors: IErrorBase[] = []
  const validate: any = field.validate
  const isValidType = Array.isArray(value)
  if (checkShouldBeRequired(field, params) && !isValidType) {
    errors.push({
        type: ErrorType.required
    })
  }
  if (isValidType) {
    if (value.length !== 2 || !isNumber(value[0]) || !isNumber(value[1])) {
      errors.push({
        type: ErrorType.invalidType
      })
    }
  }
  if (errors.length > 0) {
    return Promise.reject({mapTo: field.mapTo, errors})
  }
  return Promise.resolve(value)
}

export default validateLocation
