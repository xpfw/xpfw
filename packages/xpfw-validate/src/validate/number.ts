import { isNaN, isNil, isNumber, isString } from "lodash"
import { ErrorType } from "../globals"
import { IErrorBase, ValidateFunction } from "../typeDef"
import checkShouldBeRequired from "./util/checkShouldBeRequired"

const validateNumber: ValidateFunction = (value, field, params) => {
  const errors: IErrorBase[] = []
  const validate: any = field.validate
  let isValidType = false
  let num = value
  if (!isNil(value)) {
    if (isString(num) && num.length > 0) {
      num = Number(num)
    }
    if (!isNumber(num) || isNaN(num)) {
      errors.push({
        type: ErrorType.invalidType
      })
    } else {
      isValidType = true
    }
  }
  if (checkShouldBeRequired(field, params) && !isValidType) {
    errors.push({
        type: ErrorType.required
    })
  }
  if (isValidType && isNumber(validate.min)) {
    if (value < validate.min) {
      errors.push({
        type: ErrorType.tooLow
      })
    }
  }
  if (isValidType && isNumber(validate.max)) {
    if (value > validate.max) {
      errors.push({
        type: ErrorType.tooHigh
      })
    }
  }
  if (errors.length > 0) {
    return Promise.reject({mapTo: field.mapTo, errors})
  }
  return Promise.resolve(num)
}

export default validateNumber
