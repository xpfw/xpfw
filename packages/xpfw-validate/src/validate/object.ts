
import { each, get, isArray, isNil, isObject } from "lodash"
import { ErrorType } from "../globals"
import { IErrorBase, IField, ValidateFunction } from "../typeDef"
import validateAny from "./any"
import checkShouldBeRequired from "./util/checkShouldBeRequired"

const validateObject: ValidateFunction = (value, field, params) => {
  const errors: IErrorBase[] = []
  if (checkShouldBeRequired(field, params) && !isObject(value)) {
    errors.push({
      type: ErrorType.invalidType
    })
  }
  if (errors.length > 0) {
    return Promise.reject({mapTo: field.mapTo, errors})
  }
  const promises: Array<Promise<any>> = []
  const objectDef = get(field, `validate.objectDef`)
  if (isArray(objectDef)) {
    each(objectDef, (subFieldOptions: IField) => {
      const date = get(value, subFieldOptions.mapTo)
      promises.push(validateAny(date, subFieldOptions, params))
    })
  }
  return Promise.all(promises).then((results) => {
    return value
  })
}

export default validateObject
