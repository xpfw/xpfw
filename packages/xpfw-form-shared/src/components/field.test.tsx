import { TestDefs } from "@xpfw/validate"
import { set } from "lodash"
import * as render from "preact-render-to-json"
import * as React from "react"
import ComponentRegistry from "../componentRegistry"
import SharedField, { setValueWrap } from "../components/field"
import FormStore from "../store/form"
import baseComponentMocks, {makeMockElement} from "../testUtil/baseMock"

test("Basic Fields Test", () => {
  baseComponentMocks()
  const n: any = null
  expect(render(<SharedField field={n} />))
    .toMatchSnapshot("empty render")
  expect(render(<SharedField field={TestDefs.NameFieldPublic} />))
    .toMatchSnapshot("namefield")
  FormStore.setValue(TestDefs.NameFieldPublic.mapTo, "myval")
  expect(render(<SharedField field={TestDefs.NameFieldPublic} />))
    .toMatchSnapshot("namefield changed val")
  setValueWrap({props: {
    field: TestDefs.NameFieldPublic, form: TestDefs.FormNumberAndRequiredText
  }})("setviavaluewrap")
  expect(render((
    <SharedField
      field={TestDefs.NameFieldPublic}
      form={TestDefs.FormNumberAndRequiredText}
      theme="test"
    />
  ))).toMatchSnapshot("unfindable themed component fallsback to default")
  const newComp: any = makeMockElement("mocktest")
  ComponentRegistry.registerComponent(TestDefs.NameFieldPublic.type, newComp, "test")
  expect(render((
    <SharedField
      field={TestDefs.NameFieldPublic}
      form={TestDefs.FormNumberAndRequiredText}
      theme="test"
    />
  ))).toMatchSnapshot("use themed component")
  set(TestDefs.NameFieldPublic, "validate.defaultValue", "myDefaultTEXTS")
  FormStore.setValue(TestDefs.NameFieldPublic.mapTo, "THIS AINT THE DEFAULT VAL YO!")
  expect(render((
    <SharedField
      field={TestDefs.NameFieldPublic}
      form={TestDefs.FormNumberAndRequiredText}
      theme="test"
    />
  ))).toMatchSnapshot("render field with default component and value already set")
  FormStore.setValue(TestDefs.NameFieldPublic.mapTo, undefined)
  expect(render((
    <SharedField
      field={TestDefs.NameFieldPublic}
      form={TestDefs.FormNumberAndRequiredText}
      theme="test"
    />
  ))).toMatchSnapshot("field updated to default value because no value set")
})
