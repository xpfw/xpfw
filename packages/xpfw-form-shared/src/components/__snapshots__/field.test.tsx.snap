// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Fields Test: empty render 1`] = `
<div
  class="text"
>
  {
  "field": null
}
</div>
`;

exports[`Basic Fields Test: field updated to default value because no value set 1`] = `
<div
  class="mocktest"
>
  {
  "field": {
    "mapTo": "guestName",
    "type": 2,
    "validate": {
      "permission": {
        "find": "guest",
        "update": "guest",
        "create": "guest",
        "remove": "guest"
      },
      "defaultValue": "myDefaultTEXTS"
    }
  },
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "required": {
                "type": 2
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "theme": "test",
  "value": "myDefaultTEXTS"
}
</div>
`;

exports[`Basic Fields Test: namefield 1`] = `
<div
  class="text"
>
  {
  "field": {
    "mapTo": "guestName",
    "type": 2,
    "validate": {
      "permission": {
        "find": "guest",
        "update": "guest",
        "create": "guest",
        "remove": "guest"
      }
    }
  }
}
</div>
`;

exports[`Basic Fields Test: namefield changed val 1`] = `
<div
  class="text"
>
  {
  "field": {
    "mapTo": "guestName",
    "type": 2,
    "validate": {
      "permission": {
        "find": "guest",
        "update": "guest",
        "create": "guest",
        "remove": "guest"
      }
    }
  },
  "value": "myval"
}
</div>
`;

exports[`Basic Fields Test: render field with default component and value already set 1`] = `
<div
  class="mocktest"
>
  {
  "field": {
    "mapTo": "guestName",
    "type": 2,
    "validate": {
      "permission": {
        "find": "guest",
        "update": "guest",
        "create": "guest",
        "remove": "guest"
      },
      "defaultValue": "myDefaultTEXTS"
    }
  },
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "required": {
                "type": 2
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "theme": "test",
  "value": "THIS AINT THE DEFAULT VAL YO!"
}
</div>
`;

exports[`Basic Fields Test: unfindable themed component fallsback to default 1`] = `
<div
  class="text"
>
  {
  "field": {
    "mapTo": "guestName",
    "type": 2,
    "validate": {
      "permission": {
        "find": "guest",
        "update": "guest",
        "create": "guest",
        "remove": "guest"
      }
    }
  },
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "required": {
                "type": 2
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "theme": "test",
  "value": "setviavaluewrap"
}
</div>
`;

exports[`Basic Fields Test: use themed component 1`] = `
<div
  class="mocktest"
>
  {
  "field": {
    "mapTo": "guestName",
    "type": 2,
    "validate": {
      "permission": {
        "find": "guest",
        "update": "guest",
        "create": "guest",
        "remove": "guest"
      }
    }
  },
  "form": {
    "model": "formModel",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {
              "required": {
                "type": 2
              }
            }
          }
        ]
      },
      {
        "fields": [
          {
            "mapTo": "myNum",
            "type": 3,
            "validate": {}
          }
        ]
      }
    ]
  },
  "theme": "test",
  "value": "setviavaluewrap"
}
</div>
`;
